File: ruby/lib/weft/facilitator/fee.rb
module Weft
  module Facilitator
    module Fee
      DEFAULT_TTL = 300

      @cache = nil

      class << self
        def get_fee_info(client: Client.new, ttl: DEFAULT_TTL)
          return @cache[:fee] if cache_valid?

          supported = client.supported
          fee = supported['fee']
          raise 'Fee information not found in /supported response' unless fee

          validate_fee!(fee)

          @cache = {
            fee: fee,
            fetched_at: Time.now.to_i,
            ttl: ttl
          }

          fee
        end

        def invalidate_fee_cache
          @cache = nil
        end

        private

        def cache_valid?
          return false unless @cache

          Time.now.to_i - @cache[:fetched_at] < @cache[:ttl]
        end

        def validate_fee!(fee)
          raise 'Invalid fee structure: amount must be a string' unless fee['amount'].is_a?(String)
          raise 'Invalid fee structure: asset must be a string' unless fee['asset'].is_a?(String)
          raise 'Invalid fee structure: network must be a string' unless fee['network'].is_a?(String)
        end
      end
    end
  end
end

File: python/src/weft_sdk/facilitator/fee.py
"""
Fee information helpers for Weft Facilitator.
"""

import time
from typing import Optional

import httpx
from typing_extensions import TypedDict

from .client import WeftFacilitatorConfig, resolve_url, validate_url


class FeeInfo(TypedDict):
    amount: str
    asset: str
    network: str


class FeeCacheConfig(TypedDict, total=False):
    ttl: float


class _FeeCache(TypedDict):
    fee_info: FeeInfo
    fetched_at: float
    ttl: float


DEFAULT_CACHE_TTL = 300.0

_fee_cache: Optional[_FeeCache] = None


def invalidate_fee_cache() -> None:
    global _fee_cache
    _fee_cache = None


def _is_cache_valid() -> bool:
    if _fee_cache is None:
        return False
    now = time.time()
    return now - _fee_cache["fetched_at"] < _fee_cache["ttl"]


async def get_fee_info(
    config: Optional[WeftFacilitatorConfig] = None,
    cache_config: Optional[FeeCacheConfig] = None,
) -> FeeInfo:
    global _fee_cache

    if _is_cache_valid() and _fee_cache is not None:
        return _fee_cache["fee_info"]

    url = resolve_url(config)
    validate_url(url)

    async with httpx.AsyncClient() as client:
        response = await client.get(
            f"{url}/supported",
            headers={"Content-Type": "application/json"},
            follow_redirects=True,
        )

        if response.status_code != 200:
            raise ValueError(f"Failed to fetch fee info: {response.status_code} {response.text}")

        data = response.json()

    fee = data.get("fee")
    if fee is None:
        raise ValueError("Fee information not found in /supported response")

    if not isinstance(fee.get("amount"), str):
        raise ValueError("Invalid fee structure: amount must be a string")
    if not isinstance(fee.get("asset"), str):
        raise ValueError("Invalid fee structure: asset must be a string")
    if not isinstance(fee.get("network"), str):
        raise ValueError("Invalid fee structure: network must be a string")

    fee_info: FeeInfo = {
        "amount": fee["amount"],
        "asset": fee["asset"],
        "network": fee["network"],
    }

    ttl = cache_config.get("ttl", DEFAULT_CACHE_TTL) if cache_config else DEFAULT_CACHE_TTL
    _fee_cache = {
        "fee_info": fee_info,
        "fetched_at": time.time(),
        "ttl": ttl,
    }

    return fee_info
